File: benchmarks/http_throughput/http_throughput.rb
#!/usr/bin/env ruby
# HTTP throughput benchmark - Ruby implementation

require 'socket'

port = ARGV[0]&.to_i

if port.nil? || port == 0
  STDERR.puts "Usage: ruby http_throughput.rb <port>"
  exit 1
end

response_body = '{"message":"Hello World!"}'
response = "HTTP/1.1 200 OK\r\n" \
           "Content-Type: application/json\r\n" \
           "Content-Length: #{response_body.length}\r\n" \
           "Connection: close\r\n" \
           "\r\n" \
           "#{response_body}"

server = TCPServer.new('0.0.0.0', port)
server.setsockopt(Socket::SOL_SOCKET, Socket::SO_REUSEADDR, true)

running = true
trap('TERM') { running = false; server.close rescue nil }
trap('INT') { running = false; server.close rescue nil }

# Signal ready
puts "READY #{port}"
STDOUT.flush

while running
  begin
    client = server.accept

    # Read request line
    request = client.gets
    next if request.nil?

    # Read remaining headers
    while (line = client.gets) && line != "\r\n"
      # Skip headers
    end

    if request.start_with?('GET ')
      client.write(response)
    end

    client.close rescue nil
  rescue IOError, Errno::EBADF
    break
  rescue => e
    # Connection errors are expected during shutdown
    break unless running
  end
end

server.close rescue nil

File: benchmarks/http_throughput/http_throughput.js
#!/usr/bin/env node
// HTTP throughput benchmark - JavaScript (Node.js) implementation

const http = require('http');

const port = parseInt(process.argv[2]);

if (!port) {
    console.error('Usage: node http_throughput.js <port>');
    process.exit(1);
}

const response = JSON.stringify({ message: 'Hello World!' });
const headers = {
    'Content-Type': 'application/json',
    'Content-Length': Buffer.byteLength(response),
    'Connection': 'close'
};

const server = http.createServer((req, res) => {
    if (req.method === 'GET') {
        res.writeHead(200, headers);
        res.end(response);
    } else {
        res.writeHead(405);
        res.end();
    }
});

server.listen(port, () => {
    // Signal ready
    console.log(`READY ${port}`);
});

process.on('SIGTERM', () => {
    server.close(() => process.exit(0));
});

process.on('SIGINT', () => {
    server.close(() => process.exit(0));
});
